### PR TITLE
[Issue 6658] support multiple authz providers

### PR DESCRIPTION
**Pull request by KannarFr**
Fixes https://github.com/apache/pulsar/issues/6658

### Motivation

We are developping a custom authz provider and we don't want breaking changes on current authz usages.

### Modifications

I reproduced the way it's done for multiple authentication providers.

### Verifying this change

- [ ] Make sure that the change passes the CI checks.

This change is already covered by existing tests, such as *(please describe tests)*.

### Documentation

  - Does this pull request introduce a new feature? yes
  - If yes, how is the feature documented? docs JavaDocs

### CODE REVIEW

File: pulsar-broker-common/src/main/java/org/apache/pulsar/broker/authorization/AuthorizationService.java [hunk 1]
```diff
@@ -70,17 +90,53 @@ public AuthorizationService(ServiceConfiguration conf, ConfigurationCacheService
         }
     }
 
+    public static <T> CompletableFuture<List<T>> allAsList(List<? extends CompletionStage<? extends T>> stages) {
```
Comment (codelipenghui) [inline pulsar-broker-common/src/main/java/org/apache/pulsar/broker/authorization/AuthorizationService.java:93]: Is this method still be used?

**Comment by merlimat** [comment]
What is the semantic of multiple AuthZ providers? This is probably different in comparison with Authentication (eg: you can pass the credentials in multiple forms). 

For example, what would mean that providerA says I don't have access and providerB says I do?

**Comment by KannarFr** [comment]
The best way should be to specify authZ provider to use with authprovider. As multiple auth providers are supported, multiple authZ providers should be defined consequently.

The simplest way here, is if any authZ provider authorizes then go. 

WDYT?

**Comment by KannarFr** [comment]
The forms will the same for each authZ provider.

File: pulsar-broker-common/src/main/java/org/apache/pulsar/broker/authorization/AuthorizationService.java [hunk 2]
```diff
@@ -70,17 +90,38 @@ public AuthorizationService(ServiceConfiguration conf, ConfigurationCacheService
         }
     }
 
+    private static <T> CompletableFuture<T> anyMatch(List<? extends CompletionStage<? extends T>> l,
```
Comment (sijie) [inline pulsar-broker-common/src/main/java/org/apache/pulsar/broker/authorization/AuthorizationService.java:93]: `anyMatch` doesn't seem to be right to me when multiple authentication providers are applied to. Is it better to check them in sequence?

Comment (KannarFr) [reply pulsar-broker-common/src/main/java/org/apache/pulsar/broker/authorization/AuthorizationService.java:93]: @sijie considering https://github.com/apache/pulsar/pull/6661#issuecomment-608543559, maybe we should just define the authz provider to use on the auth provider, so dynamically use the right authz provider depending on the auth provider.

Comment (KannarFr) [reply pulsar-broker-common/src/main/java/org/apache/pulsar/broker/authorization/AuthorizationService.java:93]: So put `Set<String> getAuthZMethodsNames();` on `AuthenticationProvider` and use the defined authZ provider depending on the auth provider in `AuthorizationService`. WDYT? @sijie @merlimat

Comment (KannarFr) [reply pulsar-broker-common/src/main/java/org/apache/pulsar/broker/authorization/AuthorizationService.java:93]:
```
    /**
     * @return the authorization methods names supported by this provider
     */
    default Set<String> getAuthZMethodsNames() {
        return Sets.newHashSet("PulsarAuthorizationProvider");
    }
```

Comment (KannarFr) [reply pulsar-broker-common/src/main/java/org/apache/pulsar/broker/authorization/AuthorizationService.java:93]: The point is: how to get the right auth provider when using authZ provider methods :/.

Comment (KannarFr) [reply pulsar-broker-common/src/main/java/org/apache/pulsar/broker/authorization/AuthorizationService.java:93]: Put the auth provider method name to `AuthenticationDataSource`?

Comment (KannarFr) [reply pulsar-broker-common/src/main/java/org/apache/pulsar/broker/authorization/AuthorizationService.java:93]: Hum, or not. Anyway, I'm waiting for your point of view.

Comment (KannarFr) [reply pulsar-broker-common/src/main/java/org/apache/pulsar/broker/authorization/AuthorizationService.java:93]: Or just a simple way: as it's done for auth providers, just verify there is at least one authZ provider that authorizes the request this is why the `anyMatch`.

**Comment by KannarFr** [comment]
Close for now, the best solution would be to write an AuthorizationProvider that take a list of AuthorizationProviders. No code to change, just add a master authz provider.

**Comment by sijie** [comment]
@KannarFr cool

**Comment by KannarFr** [comment]
(Due to discuss with @sijie ;))

**Pull request closed without merging**